### PR TITLE
Warning messages for config injections in static init classes

### LINKED ISSUE #23807: Microprofile Config doesn't read config profiled properties when injected into servlet.Filter class in native mode
### Describe the bug

It seems to quarkus-undertow (processing Filter classes) doesn't work properly with profiled Quarkus configuration under **native image**.

Injected _org.eclipse.microprofile.config.Config_ instance in class implementing _javax.servlet.Filter_ doesn't contain values defined in active profile. But this happens only in native mode. Under standard JVM everything works fine and all variables (core + profiled) are available.

Filter class:
```java
@WebFilter(filterName = "BasicAuthFilter", urlPatterns = "/urlpattern/*")
public class BasicAuthFilter implements Filter {
...
 @javax.inject.Inject 
  org.eclipse.microprofile.config.Config config;

    @Override
    public void doFilter(final ServletRequest servletRequest, final ServletResponse servletResponse,
            final FilterChain filterChain) throws IOException, ServletException {
          .....
            Optional<String> env = config.getOptionalValue("stage.environment");
          .....
    }
```

Configuration file:
```yaml
quarkus:
  package:
    type: uber-jar
  log:
    level: INFO
'%NONE-CI':
  stage:
    environment: CI
```

Run under JVM:
_java -jar ... -Dquarkus.profile=NONE-CI...._

Run under Native:
_./application -Dquarkus.profile=NONE-CI_

### Expected behavior

For both, native and standard jvm, execution of _config.getOptionalValue("stage.environment")_ returns the same value if the same profile will be used.

### Actual behavior

_config.getOptionalValue("stage.environment")_ works **invalid** when application is started under **native** mode and returns empty Optional object.

_config.getOptionalValue("stage.environment")_ works properly when application is started under standard JVM mode and returns value "CI" - which is perfectly OK.

My working Workaround:
- Create some class with @ApplicationScoped (e.g. Helper)
- Put @Inject Config into this class
- Add method getConfigValue as proxy to Config instance
- Inject Helper class in my Filter class
- And use simple Helper.getConfig

```java
// request scope is also fine
@javax.enterprise.context.ApplicationScoped
public class Helper {

   // injection by field works as well
   private Config config;

    @Inject
    public void setConfig(final Config config) {
        this.config = config;
    }

   public Optional<String> getConfigValue(String key) {
         config.getOptionalValue(key);
   }
}


public .... implements javax.servlet.Filter {

   @Inject 
   Helper  helper;

   // then somewhere  in code....
     helper.getConfigValue("stage.environment");
  // returns always valid value - whether in JVM or native mode

}
```

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

Microsoft Windows [Version 10.0.19042.1466] / Native run under Docker image based on registry.access.redhat.com/ubi8/ubi-minimal:8.4

### Output of `java -version`

OpenJDK Runtime Environment Corretto-11.0.7.10.1 (build 11.0.7+10-LTS)

### GraalVM version (if different from Java)

quarkus-mandrel:21.3-java11

### Quarkus version or git rev

2.5.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T20:33:14+02:00) Java version: 11.0.7, vendor: Amazon.com Inc., runtime: C:\Programs\jdk-11-amazoncorretto Default locale: en_US, platform encoding: Cp1252 OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"

### Additional information

Application compile & build:
```
mvn clean package 
   -Pnative 
   -Dquarkus.package.type=native 
   -Dquarkus.native.container-build=true 
   -Dquarkus.native.builder-image=quay.io/quarkus/ubi-quarkus-mandrel:21.3-java11
```

I verified this also in servlet class (extends _javax.servlet.http.HttpServlet_). In this case all works properly and variables are read correctly both from directly injected Config instance and indirectly injected via Helper (application scoped) class.

### PR DESCRIPTION
**Pull request by radcortez**
- Fixes #23807

### CODE REVIEW

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java [hunk 1]
```diff
@@ -459,6 +469,67 @@ void validateConfigPropertiesInjectionPoints(
...
+    @BuildStep
+    void warnStaticInitInjectionPoints(
+            CombinedIndexBuildItem indexBuildItem,
+            ValidationPhaseBuildItem validationPhase,
+            List<ConfigClassBuildItem> configClasses,
+            BuildProducer<RunTimeConfigurationDefaultBuildItem> runTimeConfigurationDefault) {
+
+        // Add here annotated classes that are initialized during static init
+        Set<DotName> declaringClassCandidates = new HashSet<>();
+        declaringClassCandidates.add(DotName.createSimple("javax.ws.rs.ext.Provider"));
```
Comment (geoand) [inline extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java:481]: For RESTEasy Reactive, I don't think think is a problem actually...

Comment (radcortez) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java:481]: In RESTEasy Reactive, Providers are not initialized at static init? Ok, let me check and add a couple of tests there.

Comment (geoand) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java:481]: In RESTEasy Reactive they are just normal beans

Comment (radcortez) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java:481]: Confirmed. Interesting the different behavior between both. I'll add a build item.

**Review by geoand — APPROVED** [review]

**Comment by gsmet** [comment]
@radcortez I'm going through all the pull requests and unfortunately this one conflicts now. Also we would have needed a rebase anyway as CI is not conclusive enough.

**Comment by radcortez** [comment]
Done!

**Pull request merged**